metpy/calc/tools.py

```diff
@@ -976,8 +979,8 @@ def gradient(f, **kwargs):
     deprecated in 0.9 in favor of `coordinates`.
 
     """
-    pos_kwarg, positions = _process_gradient_args(kwargs)
-    return tuple(first_derivative(f, axis=ind, **{pos_kwarg: positions[ind]})
+    pos_kwarg, positions, axes = _process_gradient_args(kwargs)
+    return tuple(first_derivative(f, axis=axes[ind], **{pos_kwarg: positions[ind]})
                  for ind, pos in enumerate(positions))
```

Comment:
@dopplershift Rather than taking the derivative along all dimensions, based on how `gradient`/`laplacian` are implemented here, they only take derivatives along as many dimensions as `deltas`/`coordinates` are supplied, starting from axis 0. What these changes do is preserve this, but instead work backwards from -1 since [CF recommends](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#dimensions) that extra dimensions are leading rather than trailing (`[..., T, Z, Y, X]`), and give the option to change the order if the data do not follow that pattern.

Unless I'm misinterpreting the original intent of the design, I like this approach because, say we had some 4-D data (time, vertical, y, x). With this approach we could take both horizontal and 3-D gradients (each having their own practical uses), while effectively carrying along the extra dimension(s). In this case, if `gradient` were forced to use all dimensions, it would include time, which doesn't quite make sense to me.

In any case, I'd be glad to add some more practical tests to verify new use cases.

Reply:
You're right, that's how it ends up working, but I can't take credit for the design. The docstring for `gradient` says:
```
The first derivative calculated along each axis in the original array
```

I like your new design--let's use that explicitly in place of depending on the length of coordinates.

Reply:
Sounds good, I'll make it rely on the `axes` kwarg explicitly.

Also, just to clear some things up as I was thinking further about this, what should happen if there is a mismatch? For example

- 3-D data but only two `coordinates`/`deltas` supplied with no `axes` kwarg.
- length of `axes` kwarg not matching that of `coordinates`/`deltas`

Based on that and better understanding the original intent, I'll also plan on clearing up the docstring changes I made.

Reply:
Hmmm. Let's go with:
1. Length of coordinates/deltas should match the number of dimensions
2. Length of `axes` can be less, since that way you can request only the desired components.